nodebuilder/share/module.go

```diff
@@ -276,3 +290,105 @@ func lightAvailabilityComponents(cfg *Config) fx.Option {
 		}),
 	)
 }
+
+func archivalComponents(cfg *Config) fx.Option {
+	return fx.Options(
+		archivalDiscovery(cfg),
+		archivalPeerManager(),
+	)
+}
+
+func archivalDiscovery(cfg *Config) fx.Option {
+	return fx.Options(
+		fx.Provide(func(
+			lc fx.Lifecycle,
+			tp node.Type,
+			pruneCfg *modprune.Config,
+			d *disc.Discovery,
+			h host.Host,
+			r routing.ContentRouting,
+			opt disc.Option,
+		) (map[string]*disc.Discovery, error) {
+			// if pruner is enabled for BN, no archival service is necessary
+			if tp == node.Bridge && pruneCfg.EnableService {
+				return map[string]*disc.Discovery{"full": d}, nil
+			}
+
+			archivalDisc, err := disc.NewDiscovery(
+				cfg.Discovery,
+				h,
+				routingdisc.NewRoutingDiscovery(r),
+				archivalNodesTag,
+				opt,
+			)
+			if err != nil {
+				return nil, err
+			}
+
+			lc.Append(fx.Hook{
+				OnStart: archivalDisc.Start,
+				OnStop:  archivalDisc.Stop,
+			})
+
+			return map[string]*disc.Discovery{"full": d, "archival": archivalDisc}, nil
+		}),
+	)
+}
+
+func archivalPeerManager() fx.Option {
+	return fx.Provide(func(
+		lc fx.Lifecycle,
+		params peers.Parameters,
+		h host.Host,
+		gater *conngater.BasicConnectionGater,
+	) ([]getters.Option, disc.Option, error) {
+		archivalPeerManager, err := peers.NewManager(params, h, gater)
+		if err != nil {
+			return nil, nil, err
+		}
+		lc.Append(fx.Hook{
+			OnStart: archivalPeerManager.Start,
+			OnStop:  archivalPeerManager.Stop,
+		})
+
+		opts := []getters.Option{getters.WithArchivalPeerManager(archivalPeerManager)}
+		return opts, disc.WithOnPeersUpdate(archivalPeerManager.UpdateNodePool), nil
+	})
+}
+
+// advertiseHooks provides the lifecycle hooks for both `full` and, if
+// applicable, `archival` discovery.
+func advertiseHooks() fx.Option {
+	return fx.Invoke(func(lc fx.Lifecycle, tp node.Type, cfg *modprune.Config, discs map[string]*disc.Discovery) error {
```

Comment:
The fact that we need to hold the map in FX already suggests adding a higher-level component. Something like DiscoveryManager that has discovery CRUD (or at least CR portion for the start.) This will decrease the complexity of node builder setup and avoid hassle with FX

Reply:
will consider it